huaweicloud/services/dns/resource_huaweicloud_dns_resolver_rule_associate.go

```diff
@@ -0,0 +1,203 @@
+			Create: schema.DefaultTimeout(10 * time.Minute),
+			Delete: schema.DefaultTimeout(10 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"region": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+			"resolver_rule_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"vpc_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"status": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceDNSResolverRuleAssociateCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+	dnsClient, err := cfg.DNSV21Client(region)
+	if err != nil {
+		return diag.Errorf("error creating DNS client: %s", err)
+	}
+
+	ruleID := d.Get("resolver_rule_id").(string)
+	vpcID := d.Get("vpc_id").(string)
+	opts := associate.RouterOpts{
+		RouterID: vpcID,
+	}
+
+	body, err := associate.Associate(dnsClient, ruleID, opts).Extract()
+	if err != nil {
+		return diag.Errorf("error creating DNS resolver rule associate: %s", err)
+	}
+
+	id := fmt.Sprintf("%s/%s", ruleID, vpcID)
+	d.SetId(id)
+
+	log.Printf("[DEBUG] Waiting for DNS resolver rule associate (%s) to become available", d.Id())
+	stateConf := &resource.StateChangeConf{
+		Target:       []string{"ACTIVE"},
+		Pending:      []string{"PENDING"},
+		Refresh:      waitForDNSResolverRuleAssociate(dnsClient, ruleID, body.RouterID),
+		Timeout:      d.Timeout(schema.TimeoutCreate),
+		Delay:        5 * time.Second,
+		PollInterval: 5 * time.Second,
+	}
+
+	_, err = stateConf.WaitForStateContext(ctx)
+	if err != nil {
+		return diag.Errorf(
+			"error waiting for DNS resolver rule associate (%s) to become ACTIVE for creation: %s",
+			d.Id(), err)
+	}
+
+	return resourceDNSResolverRuleAssociateRead(ctx, d, meta)
+}
+
+func resourceDNSResolverRuleAssociateRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+	dnsClient, err := cfg.DNSV21Client(region)
+	if err != nil {
+		return diag.Errorf("error creating DNS client: %s", err)
+	}
+
+	arr := strings.Split(d.Id(), "/")
+	if len(arr) != 2 {
+		return diag.Errorf("error getting resolver rule ID and VPC ID, DNS resolver rule associate ID: %s", d.Id())
+	}
+	ruleID := arr[0]
+	vpcID := arr[1]
+
+	rule, err := resolverrule.Get(dnsClient, ruleID).Extract()
+	if err != nil {
+		return common.CheckDeletedDiag(d, err, "error retrieving DNS resolver rule")
+	}
+
+	for _, r := range rule.Routers {
+		if r.RouterID == vpcID {
+			mErr := multierror.Append(nil,
+				d.Set("vpc_id", vpcID),
+				d.Set("resolver_rule_id", ruleID),
+				d.Set("status", r.Status),
+			)
+			return diag.FromErr(mErr.ErrorOrNil())
+		}
+	}
+
+	return diag.Errorf("error retrieving DNS resolver rule associate")
```

Comment:
it's better to call CheckDeletedDiag if we can not find the VPC in all routers

Reply:
good suggestion!